### PR TITLE
GEOMETRY-115: IO Modules

### PR DESCRIPTION
**Pull request by darkma773r**
Adding modules
- ~~commons-geometry-core-io~~ commons-geometry-io-core
- ~~commons-geometry-euclidean-io~~ commons-geometry-euclidean-io

### CODE REVIEW

File: commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/BoundaryReadHandler3D.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * be thread-safe.</p>
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Boundary_representations">Boundary representations</a>
+ * @see BoundaryWriteHandler3D
+ * @see BoundaryIOManager3D
+ */
+public interface BoundaryReadHandler3D extends BoundaryReadHandler<PlaneConvexSubset, BoundarySource3D> {
+
+    /** Return a {@link FacetDefinitionReader} for reading raw
+     * {@link org.apache.commons.geometry.euclidean.io.threed.facet.FacetDefinition facets} from the given
```
Comment (arturobernalg) [inline commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/BoundaryReadHandler3D.java:44]: Fix javadoc. --> org.apache.commons.geometry.euclidean.io.threed.FacetDefinition

File: commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/txt/TextBoundaryWriteHandler3D.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+import java.text.DecimalFormat;
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+import org.apache.commons.geometry.core.io.internal.GeometryIOUtils;
+import org.apache.commons.geometry.euclidean.io.threed.AbstractBoundaryWriteHandler3D;
+import org.apache.commons.geometry.euclidean.io.threed.FacetDefinition;
+import org.apache.commons.geometry.euclidean.threed.PlaneConvexSubset;
+
+/** {@link BoundaryWriteHandler3D} implementation designed to write simple text data
```
Comment (arturobernalg) [inline commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/txt/TextBoundaryWriteHandler3D.java:31]: fix javadoc org.apache.commons.geometry.euclidean.io.threed.BoundaryWriteHandler3D

File: commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/txt/TextBoundaryWriteHandler3D.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+ * formats using {@link TextFacetDefinitionWriter}. Output is written using the UTF-8 charset
+ * by default.
+ * @see BoundaryWriteHandler3D
```
Comment (arturobernalg) [inline commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/txt/TextBoundaryWriteHandler3D.java:34]: idem

File: commons-geometry-core-io/src/main/java/org/apache/commons/geometry/core/io/internal/SimpleTextParser.java [hunk 4]
```diff
@@ -0,0 +1,1207 @@
...
+    private String currentToken;
+
+    /** The line number that the current token started on. */
+    private int currentTokenLineNumber = INITIAL_TOKEN_POS;
+
+    /** The character number that the current token started on. */
+    private int currentTokenColumnNumber = INITIAL_TOKEN_POS;
+
+    /** Flag used to indicate that at least one token has been read from the stream. */
+    private boolean hasSetToken = false;
```
Comment (arturobernalg) [inline commons-geometry-core-io/src/main/java/org/apache/commons/geometry/core/io/internal/SimpleTextParser.java:73]: we can safely remove the redundant initializer

File: commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/AbstractBoundaryReadHandler3D.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        }
+
+        /** Load the next facet from the underlying reader. Any {@link IOException}s
+         * are wrapped with {@link java.io.UncheckedIOException}.
+         */
+        private void loadNext() {
+            ++loadCount;
+            try {
+                next = reader.readFacet();
+            } catch (IOException exc) {
```
Comment (arturobernalg) [inline commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/AbstractBoundaryReadHandler3D.java:160]: We can use final

File: commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/obj/OBJTriangleMeshReader.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        PolygonOBJParser.Face face;
+        Vector3D definedNormal;
+        Iterator<PolygonOBJParser.VertexAttributes> attrs;
+        while ((face = readFace()) != null) {
+            // get the face attributes in the proper counter-clockwise orientation
+            definedNormal = face.getDefinedCompositeNormal(normals::get);
+            attrs = face.getCounterClockwiseVertexAttributes(definedNormal, meshBuilder::getVertex).iterator();
+
+            // add the face vertices using a triangle fan
+            int p0 = attrs.next().getVertexIndex();
```
Comment (arturobernalg) [inline commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/obj/OBJTriangleMeshReader.java:67]: we can use final

File: commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/txt/TextFacetDefinitionReader.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+     * @throws IOException if an I/O or parsing error occurs
+     */
+    private FacetDefinition readFacetInternal() throws IOException {
+        final Vector3D p1 = readVector();
+        discardNonData();
+        final Vector3D p2 = readVector();
+        discardNonData();
+        final Vector3D p3 = readVector();
+
+        List<Vector3D> vertices;
```
Comment (arturobernalg) [inline commons-geometry-euclidean-io/src/main/java/org/apache/commons/geometry/euclidean/io/threed/txt/TextFacetDefinitionReader.java:169]: We can use final

**Review by arturobernalg — CHANGES_REQUESTED** [review]
IMO It's importan fix the javadoc. Thank you for the PR

**Comment by darkma773r** [comment]
Thank you for the detailed review, @arturobernalg!

**Comment by arturobernalg** [comment]
> Thank you for the detailed review, @arturobernalg!

Great. Glad to help @darkma773r. Please let me know If I can help you white anything.

**Comment by darkma773r** [comment]
Renamed modules and packages to have `io` portion first in order to avoid JPMS conflicts. (Based on discussion on dev mailing list.)

**Pull request closed without merging**